### PR TITLE
[Relay][Compilation] replace relay.build_module with C++ BuildModule

### PR DESCRIPTION
**Pull request by zhiics**
This PR ports the relay.build_module from Python to using C++ APIs developed in #3082. Some Python codes are still needed, like AutoTVM related ones. In addition, some fp16 tests are added.

@antinucleon @jroesch @wweic @MarisaKirisame @tqchen

### CODE REVIEW

File: python/tvm/relay/backend/graph_runtime_codegen.py [hunk 1]
```diff
@@ -57,17 +54,14 @@ def __init__(self, mod, target):
         self._setup(mod, target)
 
     def _setup(self, mod, target):
-        tgts = []
+        tgts = {}
         if isinstance(target, dict):
             for kv in target.items():
```
Comment (wweic) [inline python/tvm/relay/backend/graph_runtime_codegen.py]:
```suggestion
            for device, target in target.items():
```

Comment (zhiics) [reply python/tvm/relay/backend/graph_runtime_codegen.py]: This is from the old commit. But I can change it anyway.

File: python/tvm/relay/build_module.py [hunk 2]
```diff
@@ -129,108 +103,176 @@ def build_config(**kwargs):
...
-    params : Optional[Dict[str, tvm.nd.NDArray]]
-        Input parameters to the graph that do not change
-        during inference time. used for constant folding.
 
-    Returns
-    -------
-    opt_func : tvm.relay.Function
-        The optimized version of the function.
+class BuildModule(object):
+    """Build a Relay function to run on TVM graph runtime.
```
Comment (tqchen) [inline python/tvm/relay/build_module.py]: document a bit more. Module class for exposing relay.build in c++

File: python/tvm/relay/build_module.py [hunk 3]
```diff
@@ -129,108 +103,176 @@ def build_config(**kwargs):
...
+    def get_params(self):
+        """Return the updated weights."""
+        params = self._get_params_func()
+        ret = {}
+        for key, value in params.items():
+            ret[key] = value.data
+        return ret
+
+    def set_opt_level(self, level):
+        """Set the optimization level."""
```
Comment (tqchen) [inline python/tvm/relay/build_module.py]: document all the arguments

File: python/tvm/relay/build_module.py [hunk 4]
```diff
@@ -129,108 +103,176 @@ def build_config(**kwargs):
...
+                param = _nd.array(param)
+            inputs[name] = _expr.const(param)
+        self._set_params_func(inputs)
+
+    def add_pass(self, pass_name):
+        """Add a pass to the pass list"""
+        self._add_pass(pass_name)
+
+    def disable_pass(self, pass_name):
+        """Add a pass to the disabled pass list"""
```
Comment (tqchen) [inline python/tvm/relay/build_module.py]: document all the arguments

File: src/relay/backend/build_module.cc [hunk 5]
```diff
@@ -38,26 +38,23 @@ namespace tvm {
...
-  static int Str2Mask(const std::string& str) {
-    CHECK_GT(str2mask.count(str), 0) << "Unknown context.";
-    return str2mask.at(str);
-  }
 };
 
 const std::unordered_map<int, std::string> ContextMap::mask2str = {
-  {1, "cpu"},
-  {2, "gpu"},
+  {1, "llvm"},
```
Comment (tqchen) [inline src/relay/backend/build_module.cc]: The original mapping seems to be correct, any reason for this change?

**Comment by tqchen** [comment]
Need a few more reviews. @jroesch can you manage this PR?

File: src/relay/backend/build_module.cc (hunk 5 above)
Comment (zhiics) [reply src/relay/backend/build_module.cc]: We actually need to create `Target` from device type here: https://github.com/dmlc/tvm/blob/master/src/relay/backend/build_module.cc#L599. So I changed it from context name to target string. I can probably change the variable name to `ContextTargetMap` as well.

File: python/tvm/relay/build_module.py [hunk 6]
```diff
@@ -129,108 +103,203 @@ def build_config(**kwargs):
...
-        if k not in name_dict:
-            continue
-        arg = name_dict[k]
-        if arg is None:
-            raise ValueError("Multiple args in the function have name %s" % k)
-        bind_dict[arg] = _expr.const(v)
-    return _expr.bind(func, bind_dict)
-
-
-def optimize(func, target=None, params=None):
```
Comment (vinx13) [inline python/tvm/relay/build_module.py]: `optimize` is used in https://github.com/dmlc/tvm/blob/6a4d71ff40915611bd42b62994992b879e6be610/python/tvm/relay/quantize/quantize.py#L300
can you fix it?
I noticed that for some reason e2e quantization test is disabled.

Comment (zhiics) [reply python/tvm/relay/build_module.py]: @vinx13 Thanks for pointing out. For quantization test, it was tracked here by #2718 

I can probably enable it in a different PR.

**Comment by jroesch** [comment]
> Need a few more reviews. @jroesch can you manage this PR?

Yes I can, I was waiting for everyone to take some review passes.

cc @antinucleon can you take a look as well?

File: tests/python/relay/test_cpp_build_module.py [hunk 7]
```diff
@@ -86,21 +44,61 @@ def test_build():
...
-        np.maximum(np.dot(A.asnumpy(), B.asnumpy().T), 0) + C.asnumpy(), atol=1e-5, rtol=1e-5)
-  
+
+    np.testing.assert_allclose(out.asnumpy(), np.maximum(np.dot(A.asnumpy(),
+                                                                B.asnumpy().T),
+                                                         0) + C.asnumpy(),
+                               atol=1e-5, rtol=1e-5)
+
+
+def test_fp16_build():
```
Comment (hlu1) [inline tests/python/relay/test_cpp_build_module.py]: Can you add fp16 test for CPU as well? Intel CPUs do not support fp16 compute, but they do support fp16 as storage format.

File: python/tvm/relay/build_module.py [hunk 8]
```diff
@@ -129,108 +103,203 @@ def build_config(**kwargs):
...
+        target : str, :any:`tvm.target.Target`, or dict of str(i.e.
+        device/context name) to str/tvm.target.Target, optional
+            For heterogeneous compilation, it is a dictionary indicating context
+            to target mapping. For homogeneous compilation, it is a build target.
+
+        target_host : str or :any:`tvm.target.Target`, optional
+            Host compilation target, if target is device.
+            When TVM compiles device specific program such as CUDA,
+            we also need host(CPU) side code to interact with the driver
+            setup the dimensions and parameters correctly.
```
Comment (hlu1) [inline python/tvm/relay/build_module.py]: Nit: `to setup the dimensions and parameters correctly.`

File: python/tvm/relay/build_module.py (hunk 6 above)
Comment (zhiics) [reply python/tvm/relay/build_module.py]: I ended up enabling the quantization tests in this PR as we need to update the `optimize` function here. I will keep my eye on it and ping @ZihengJiang once it's flaky again.

File: tests/python/relay/test_cpp_build_module.py (hunk 7 above)
Comment (zhiics) [reply tests/python/relay/test_cpp_build_module.py]: Added, plz double check.

File: src/relay/backend/build_module.cc [hunk 9]
```diff
@@ -530,25 +495,19 @@ class RelayBuildModule : public runtime::ModuleNode {
...
-    if (tmp_map.count(fallback_idx) == 0) {
-      tmp_map[fallback_idx] = cfg.fallback_device;
-    }
-    for (const auto& kv : tmp_map) {
+    if (tmp_map.count(cfg.fallback_device) == 0) {
       device_target.Set(
-        ir::IntImm::make(HalideIR::Int(64), kv.first),
-        ir::StringImm::make(kv.second));
+          cfg.fallback_device,
+          Target::create(ContextTargetMap::Mask2Str(cfg.fallback_device)));
```
Comment (jroesch) [inline src/relay/backend/build_module.cc]: See my previous comment, why do we need to allocate a target

File: src/relay/backend/build_module.cc [hunk 10]
```diff
@@ -69,25 +65,6 @@ const std::unordered_map<int, std::string> ContextMap::mask2str = {
   {12, "ext_dev"}
 };
 
-const std::unordered_map<std::string, int> ContextMap::str2mask = {
```
[the remark this answers is not in the document]
Comment (zhiics) [reply src/relay/backend/build_module.cc]: We passed in device type. Therefore, we don't really need this.

File: src/relay/backend/build_module.cc [hunk 11]
```diff
@@ -38,26 +38,22 @@ namespace tvm {
...
+using TargetsMap = Map<tvm::Integer, tvm::Target>;
+
 /*!
- * \brief Context name / index
- *        See: python/tvm/_ffi/runtime_ctypes.py
+ * \brief Context index to Target
  */
-struct ContextMap {
+struct ContextTargetMap {
   static const std::unordered_map<int, std::string> mask2str;
```
[the remark this answers is not in the document]
Comment (zhiics) [reply src/relay/backend/build_module.cc]: We can, but we them need to create the target objects here. I don't think that's quite necessary. @antinucleon, right?

[the remark this answers is not in the document]
Comment (antinucleon) [reply src/relay/backend/build_module.cc]: I keep same to Python version. In python graph codegen there are a lot of place requires string target. With a refactor we can remove all these string.

**Comment by tqchen** [comment]
Thanks, @zhiics @jroesch @vinx13 @wweic @hlu1 this PR is now merged

**Comment by kparzysz** [comment]
This results in a runtime message: `src/codegen/build_module.cc:132: Unknown target name vpi`.

See [https://discuss.tvm.ai/t/unknown-target-name-vpi/2589](https://discuss.tvm.ai/t/unknown-target-name-vpi/2589).

**Comment by tqchen** [comment]
https://github.com/dmlc/tvm/issues/3197

**Pull request merged**